### PR TITLE
Add Service account support and refactor credentials and oauth logic.

### PR DESCRIPTION
**Pull request by BenRKarl**
**NOTE**: This change includes breaking changes listed below. It should be released along a major version boundary:

* The `client._get_client_kwargs_from_env` method is removed.
* The `client._get_client_kwargs_from_yaml` method is removed.
* The `client._validate_login_customer_id` method is moved to `google.ads.google_ads.config.validate_login_customer_id`.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -192,6 +212,8 @@ Authors
 * `David Wihl`_
 * `Ben Karl`_
 
+.. _Installed Application Flow: https://developers.google.com/google-ads/api/docs/client-libs/dotnet/oauth-installed
```
Comment (wihl) [inline README.rst]: this is pointing to the .NET page. Without moving over the whole wiki, can we at least add the page for Python? Same comment for the following line.

File: google-ads.yaml [hunk 2]
```diff
@@ -1,14 +1,37 @@
...
 login_customer_id: INSERT_LOGIN_CUSTOMER_ID_HERE
 
+-# Service Account configuration
+-###############################################################################
+-# To authenticate with a service account add the appropriate values to the    #
+-# below configuration parameters and remove the four OAuth credentials above. #
+-# The "path_to_private_key_file" value should be a path to your local private #
+-# key json file, and "delegated_account" should be the email address that is  #
+-# being used to impersonate the credentials making requests. for more         #
+-# information on service accounts, see:                                       #
```
Comment (wihl) [inline google-ads.yaml]: might want to add a pointer to the README about removing the Installed Application Flow in order to use these parameter.

File: google/ads/google_ads/client.py [hunk 3]
```diff
@@ -214,16 +113,34 @@ def load_from_storage(cls, path=None):
...
+    def get_type(cls, name, version=_DEFAULT_VERSION):
+        """Returns the specified common, enum, error, or resource type.
 
-        with open(path, 'rb') as handle:
-            yaml_str = handle.read()
+        Args:
+            name: a str indicating the name of the type that is being retrieved;
+                e.g. you may specify "CampaignOperation" to retrieve a
+                CampaignOperation instance.
+            version: a str indicating the version of the Google Ads API to be
```
Comment (wihl) [inline google/ads/google_ads/client.py]: Nit: remove "to be used" so it is cleaner and fits on one line

File: google/ads/google_ads/client.py [hunk 4]
```diff
@@ -214,16 +113,34 @@ def load_from_storage(cls, path=None):
...
+            A Message instance representing the desired type.
+
+        Raises:
+            AttributeError: If the type for the specified name doesn't exist
+                in the given version.
+        """
+        try:
+            message_type = getattr(_get_version(version).types, name)
+        except AttributeError:
+            raise ValueError('Specified type "%s" does not exist in Google Ads '
```
Comment (wihl) [inline google/ads/google_ads/client.py]: .format()?

File: google/ads/google_ads/config.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    list(_OPTIONAL_KEYS) +
+    list(_OAUTH2_INSTALLED_APP_KEYS) +
+    list(_OAUTH2_SERVICE_ACCOUNT_KEYS)}
+
+
+def _config_validation_decorator(func):
+    """A decorator used to easily run validations on configs loaded into dicts.
+
+    Add this decorator to any new methods in this module that read library
+    configuration and return it as a dict.
```
Comment (wihl) [inline google/ads/google_ads/config.py]: The grammar is a little wonky. "Add this decorator to any method that returns the config as a dict."

File: google/ads/google_ads/oauth2.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import functools
+
+from google.oauth2.service_account import Credentials as ServiceAccountCreds
+from google.oauth2.credentials import Credentials as InstalledAppCredentials
+from google.auth.transport.requests import Request
+
+from google.ads.google_ads import config
+
+_SERVICE_ACCOUNT_SCOPES = ['https://www.googleapis.com/auth/adwords']
+_DEFAULT_TOKEN_URI = 'https://accounts.google.com/o/oauth2/token'
```
Comment (wihl) [inline google/ads/google_ads/oauth2.py]: Endpoint may have changed. See https://github.com/golang/oauth2/issues/310

File: google/ads/google_ads/oauth2.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            config_data.get('client_id'),
+            config_data.get('client_secret'),
+            config_data.get('refresh_token'))
+    elif all(key in config_data for key in required_service_account_keys):
+        # Using the Service Account Flow
+        return get_service_account_credentials(
+            config_data.get('path_to_private_key_file'),
+            config_data.get('delegated_account'))
+    else:
+        raise ValueError('Your yaml file is incorrectly configured for OAuth2. '
```
Comment (wihl) [inline google/ads/google_ads/oauth2.py]: Nit: s/yaml/YAML

File: google/ads/google_ads/oauth2.py (hunk 6 above)
Comment (BenRKarl) [reply google/ads/google_ads/oauth2.py]: Both URLs seem to work, let's investigate further whether a change is necessary.

Comment (BenRKarl) [reply google/ads/google_ads/oauth2.py]: We couldn't find further evidence this is an issue so I'm resolving.

File: README.rst (hunk 1 above)
Comment (BenRKarl) [reply README.rst]: Thanks for catching this! I still need to add the Python docs to that page. Let's discuss if it makes sense to add just the sections linked here.

File: google/ads/google_ads/client.py (hunk 3 above)
Comment (BenRKarl) [reply google/ads/google_ads/client.py]: Done

File: google/ads/google_ads/client.py (hunk 4 above)
Comment (BenRKarl) [reply google/ads/google_ads/client.py]: Done

File: google/ads/google_ads/config.py (hunk 5 above)
Comment (BenRKarl) [reply google/ads/google_ads/config.py]: Done

File: google/ads/google_ads/oauth2.py (hunk 7 above)
Comment (BenRKarl) [reply google/ads/google_ads/oauth2.py]: Done

File: google-ads.yaml (hunk 2 above)
Comment (BenRKarl) [reply google-ads.yaml]: Done

File: README.rst (hunk 1 above)
Comment (BenRKarl) [reply README.rst]: This is updated now that the Python devsite section is published.

**Comment by BenRKarl** [comment]
@akaashhazarika I don't think that will work as expected because the sets may not be equal, it's possible for config data to have more than just the required installed app keys. What that validation cares about is whether all of the required installed app keys are included in the config data keys.

**Comment by wihl** [comment]
LGTM

**Pull request merged**